src/utils/auth.js

```diff
@@ -22,51 +22,55 @@ async function withToken(cb) {
   cb(await auth0.getTokenSilently());
 }
 
+async function kickOffAuthentication() {
+  // Auth0 requires that the user be redirected to a true static path
+  // after authentication.[1] In our case, the only true static path is
+  // the root path. Everything else is handled by the client-side router.
+  //
+  // However, we can still record the path that the user attempted to
+  // access so that we can return them to it after they authenticate
+  // (above).
+  //
+  // [1] https://bugzilla.mozilla.org/show_bug.cgi?id=1623800#c1
+  sessionStorage.setItem('realAuthRedirectTarget', window.location.href);
+  await auth0.loginWithRedirect({
+    redirect_uri: window.location.origin,
+  });
+}
+
+async function wrapUpAuthentication() {
+  await auth0.handleRedirectCallback();
+  const realAuthRedirectTarget = sessionStorage.getItem(
+    'realAuthRedirectTarget'
+  );
+  sessionStorage.removeItem('realAuthRedirectTarget');
+
+  // It's important to use window.location.replace here rather than
+  // window.history.replaceState or window.history.pushState because we
+  // want the app to load from scratch once the user arrives at the target
+  // path. That way, the store will be initialized based on the target
+  // path.
+  //
+  // Using window.location.replace also ensures that the homepage does not
+  // appear in the user's history between auth0.com and the target path.
+  window.location.replace(realAuthRedirectTarget);
+}
+
 export function authenticate(successCallback) {
   window.onload = async () => {
     await configureClient();
 
     const isAuthenticated = await auth0.isAuthenticated();
+    const codeAndStatePresent =
+      window.location.search.includes('code=') &&
```

Comment:
good candidate for a util function, we repeat this in 2 places

Reply:
Good call!